core/src/main/java/org/apache/iceberg/PartitionStatsUtil.java

```diff
@@ -0,0 +1,136 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Queues;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.types.Types.StructType;
+import org.apache.iceberg.util.Pair;
+import org.apache.iceberg.util.PartitionMap;
+import org.apache.iceberg.util.PartitionUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.iceberg.util.ThreadPools;
+
+public class PartitionStatsUtil {
+
+  private PartitionStatsUtil() {}
+
+  /**
+   * Computes the partition stats for the given snapshot of the table.
+   *
+   * @param table the table for which partition stats to be computed.
+   * @param snapshot the snapshot for which partition stats is computed.
+   * @return the collection of {@link PartitionStats}
+   */
+  public static Collection<PartitionStats> computeStats(Table table, Snapshot snapshot) {
+    Preconditions.checkArgument(table != null, "table cannot be null");
+    Preconditions.checkArgument(snapshot != null, "snapshot cannot be null");
+
+    StructType partitionType = Partitioning.partitionType(table);
+    if (partitionType.fields().isEmpty()) {
+      throw new UnsupportedOperationException(
+          "Computing partition stats for an unpartitioned table");
+    }
+
+    List<ManifestFile> manifests = snapshot.allManifests(table.io());
+    Queue<PartitionMap<PartitionStats>> statsByManifest = Queues.newConcurrentLinkedQueue();
+    Tasks.foreach(manifests)
+        .stopOnFailure()
+        .throwFailureWhenFinished()
+        .executeWith(ThreadPools.getWorkerPool())
+        .run(manifest -> statsByManifest.add(collectStats(table, manifest, partitionType)));
+
+    return mergeStats(statsByManifest, table.specs());
+  }
+
+  /**
+   * Sorts the {@link PartitionStats} based on the partition data.
+   *
+   * @param stats collection of {@link PartitionStats} which needs to be sorted.
+   * @param partitionType unified partition schema.
+   * @return the list of {@link PartitionStats}
+   */
+  public static List<PartitionStats> sortStats(
+      Collection<PartitionStats> stats, StructType partitionType) {
+    List<PartitionStats> entries = Lists.newArrayList(stats.iterator());
+    entries.sort(
+        Comparator.comparing(PartitionStats::partition, Comparators.forType(partitionType)));
+    return entries;
+  }
+
+  private static PartitionMap<PartitionStats> collectStats(
+      Table table, ManifestFile manifest, StructType partitionType) {
+    try (ManifestReader<?> reader = openManifest(table, manifest)) {
+      PartitionMap<PartitionStats> statsMap = PartitionMap.create(table.specs());
+
+      for (ManifestEntry<?> entry : reader.entries()) {
+        ContentFile<?> file = entry.file();
+        PartitionSpec spec = table.specs().get(file.specId());
+        int specId = file.specId();
+        PartitionData partition =
+            PartitionUtil.coercePartitionData(partitionType, spec, file.partition());
```

Comment:
This may cause unnecessary performance overhead. Take a look at [this](https://github.com/apache/iceberg/pull/9629) PR. If I am not mistaken (please, double check), we should do something like this.

```
PartitionData keyTemplate = new PartitionData(partitionType);

for (ManifestEntry<?> entry : reader.entries()) {
  ContentFile<?> file = entry.file();
  StructLike partition = file.partition();
  StructLike coercedPartition = PartitionUtil.coercePartition(partitionType, spec, partition);
  StructLike key = keyTemplate.copyFor(coercedPartition);
  Snapshot snapshot = table.snapshot(entry.snapshotId());
  PartitionStats stats =
      statsMap.computeIfAbsent(specId, key, () -> new PartitionStats(key, specId));
  if (entry.isLive()) {
    stats.liveEntry(file, snapshot);
  } else {
    stats.deletedEntry(snapshot);
  }
}
```

Using `keyTemplate.copyFor` should avoid the overhead described [here](https://github.com/apache/iceberg/pull/9629#discussion_r1476890643). Let me know if this makes sense.